proposals/NNNN-objc-implementation.md

```diff
@@ -0,0 +1,323 @@
+
+## Detailed design
+
+### Custom category names
+
+As an enabling step, we propose that using `@objc(CustomName)` on an `extension` should cause the Objective-C category created for that extension to be named `CustomName`, rather than using a category name generated by the compiler. This name should appear in both generated headers and Objective-C metadata. The compiler should enforce that there is only one extension per class/category-name combination.
+
+`@objc(CustomName) extension` also has the effect of `@objc extension`, namely, making members of the extension `@objc` by default.
+
+### `@implementation`
+
+The compiler will accept a new attribute, `@implementation`, which turns a declaration that would normally be exported to another language into a declaration that implements something imported from another language. This attribute takes no arguments; it should be used alongside another attribute specifying the foreign language, such as `@objc`.
+
+In general, `@implementation` will cause the affected declarations to be matched against imported declarations, and will emit errors if a good enough match cannot be found. It cannot necessarily be applied to all declarations that can be exported to a foreign language; for instance, `@objc @implementation` is only supported on whole extensions, not on individual members.
+
+### `@objc @implementation extension`s
+
+`@objc @implementation extension` causes an extension to be used as the implementation of an Objective-C `@interface` declaration for the class it extends. If the `@objc` attribute specifies a custom category name, it will implement a category with that name; otherwise it will implement the main `@interface` for the class (plus any class extensions).
+
+```swift
+@objc @implementation extension SomeClass {
+    // Equivalent to `@implementation SomeClass`;
+    // implements everything in `@interface SomeClass` and
+    // all `@interface SomeClass ()` extensions.
+}
+
+@objc(SomeCategory) @implementation extension SomeClass {
+    // Equivalent to `@implementation SomeClass (SomeCategory)`;
+    // implements everything in `@interface SomeClass (SomeCategory)`.
+}
+```
+
+As in any `@objc extension`, all members are implicitly `@objc` by default, and all `@objc` members are implicitly `dynamic`; unlike other `@objc extension`s, adding the `final` keyword makes the declaration *not* `@objc` by default.
+
+As a special exception to the usual rule, an `@objc @implementation extension` which implements the main Objective-C interface of a class can declare stored properties, designated and required `init`s, and `deinit`s.
+
+#### Rules
+
+An `@objc @implementation extension` must:
+
+* Extend a non-root class imported from Objective-C which does not use lightweight generics.
```

Comment:
Is this really 'non-root'? Or is it 'subclass of `NSObject`?

Reply:
The rule the compiler enforces is "non-root". This is the same rule enforced for `@objc` classes—the type checker permits you to subclass `NSProxy`, for instance. (I can't guarantee that these subclasses *work*, although I don't know of a reason why they wouldn't.)